src/Compilers/Core/Portable/Operations/OperationExtensions.cs

```diff
@@ -1,13 +1,76 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Threading;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Semantics
 {
     public static class OperationExtensions
     {
+        /// <summary>
+        /// Find the argument supplied for a given parameter of the target method.
+        /// </summary>
+        /// <param name="hasArgumentsExpression">The IHasArgumentsExpression object to get matching argument object from.</param>
+        /// <param name="parameter">Parameter of the target method.</param>
+        /// <returns>Argument corresponding to the parameter.</returns>
+        public static IArgument GetArgumentMatchingParameter(this IHasArgumentsExpression hasArgumentsExpression, IParameterSymbol parameter)
+        {
+            if (hasArgumentsExpression == null || parameter == null)
+            {
+                return null;
+            }
+
+            foreach (var argument in hasArgumentsExpression.ArgumentsInEvaluationOrder)
+            {
+                if (argument.Parameter == parameter)
+                {
+                    return argument;
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Arguments of the invocation, excluding the instance argument. Arguments are in parameter order,
+        /// and params/ParamArray arguments have been collected into arrays. Default values are supplied for
+        /// optional arguments missing in source.
+        /// </summary>
+        public static ImmutableArray<IArgument> ArgumentsInParameterOrder(this IHasArgumentsExpression hasArgumentsExpression)
```

Comment:
The concerns you have regarding these extension methods depend on how are we gonna expose invocation and arguments in erroneous code. Like we discussed earlier, we might need a new operation to have a consistent representation for this. I will file an interface design issue for that.

Consider this is not part of the core IOperation interface, I will probably removed these extension methods from this PR for now, and file an issue to implement them after the design is decided. Does it sounds reasonable?